Quarkus Spring Data DTO and Projections support

**Describe the bug**
Spring Data API supports DTO and Projections.
Here you can read how it should work:
https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#projections

**Expected behavior**
DTO and Projections should work

**Actual behavior**
It just try to retrieve the full object and fails with Cast error:
```
Caused by: java.lang.ClassCastException: class org.acme.entities.Contact cannot be cast to class org.acme.data.SimpleContactProjection (org.acme.entities.Contact and org.acme.data.SimpleContactProjection are in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @363f6148)
```


**To Reproduce**
Please check this repo:
https://github.com/engilyin/quarkus-spring-dto.git

All tests should be passed (build succeed):
./gradlew clean build

Also you can find commented out stuff in this repo. That was my try to use dynamic projections/DTO.
It breaks build at all.

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: Darwin MacBook-Pro-Alexander.local 19.6.0 Darwin Kernel Version 19.6.0: Mon Aug 31 22:12:52 PDT 2020; root:xnu-6153.141.2~1/RELEASE_X86_64 x86_64

 - Output of `java -version`:  openjdk 11.0.5 2019-10-15 LTS
 - Quarkus version or git rev: 1.9.1.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): Gradle 6.6.1

--- a/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/AbstractMethodsAdder.java
+++ b/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/AbstractMethodsAdder.java
@@ -16,6 +16,8 @@ import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationValue;
 import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
+import org.jboss.jandex.IndexView;
+import org.jboss.jandex.Type;
 import org.springframework.data.domain.PageImpl;
 import org.springframework.data.domain.Pageable;
 import org.springframework.data.domain.SliceImpl;
@@ -31,6 +33,7 @@ import io.quarkus.gizmo.ResultHandle;
 import io.quarkus.gizmo.TryBlock;
 import io.quarkus.hibernate.orm.panache.PanacheQuery;
 import io.quarkus.panache.common.Page;
+import io.quarkus.runtime.util.HashUtil;
 import io.quarkus.spring.data.deployment.DotNames;
 import io.quarkus.spring.data.runtime.RepositorySupport;
 import io.quarkus.spring.data.runtime.TypesConverter;
@@ -66,7 +69,7 @@ public abstract class AbstractMethodsAdder {
 
     protected void generateFindQueryResultHandling(MethodCreator methodCreator, ResultHandle panacheQuery,
             Integer pageableParameterIndex, ClassInfo repositoryClassInfo, ClassInfo entityClassInfo,
-            DotName returnType, Integer limit, String methodName, DotName customResultType) {
+            DotName returnType, Integer limit, String methodName, DotName customResultType, String originalResultType) {
 
         ResultHandle page = null;
         if (limit != null) {
@@ -128,7 +131,7 @@ public abstract class AbstractMethodsAdder {
                 ResultHandle customResult = tryBlock.invokeStaticMethod(
                         MethodDescriptor.ofMethod(customResultType.toString(), "convert_" + methodName,
                                 customResultType.toString(),
-                                Object[].class.getName()),
+                                originalResultType),
                         singleResult);
                 ResultHandle optional = tryBlock.invokeStaticMethod(
                         MethodDescriptor.ofMethod(Optional.class, "of", Optional.class, Object.class),
@@ -153,13 +156,14 @@ public abstract class AbstractMethodsAdder {
                         MethodDescriptor.ofMethod(PanacheQuery.class, "stream", Stream.class),
                         panacheQuery);
 
-                // Function to convert Object[] to the custom type (using the generated static convert method)
+                // Function to convert `originResultType` (Object[] or entity class)
+                // to the custom type (using the generated static convert method)
                 FunctionCreator customResultMappingFunction = methodCreator.createFunction(Function.class);
                 BytecodeCreator funcBytecode = customResultMappingFunction.getBytecode();
                 ResultHandle obj = funcBytecode.invokeStaticMethod(
                         MethodDescriptor.ofMethod(customResultType.toString(), "convert_" + methodName,
                                 customResultType.toString(),
-                                Object[].class.getName()),
+                                originalResultType),
                         funcBytecode.getMethodParam(0));
                 funcBytecode.returnValue(obj);
 
@@ -245,7 +249,7 @@ public abstract class AbstractMethodsAdder {
             ResultHandle customResult = tryBlock.invokeStaticMethod(
                     MethodDescriptor.ofMethod(customResultType.toString(), "convert_" + methodName,
                             customResultType.toString(),
-                            Object[].class.getName()),
+                            originalResultType),
                     singleResult);
 
             tryBlock.returnValue(customResult);
@@ -293,5 +297,57 @@ public abstract class AbstractMethodsAdder {
 
     protected boolean isHibernateSupportedReturnType(DotName dotName) {
         return dotName.equals(DotNames.OBJECT) || DotNames.HIBERNATE_PROVIDED_BASIC_TYPES.contains(dotName);
+    }
+
+    protected Type verifyQueryResultType(Type t, IndexView index) {
+        if (isHibernateSupportedReturnType(t.name())) {
+            return t;
+        }
+        if (t.kind() == Type.Kind.ARRAY) {
+            return verifyQueryResultType(t.asArrayType().component(), index);
+        } else if (t.kind() == Type.Kind.PARAMETERIZED_TYPE) {
+            final List<Type> types = t.asParameterizedType().arguments();
+            if (types.size() == 1) {
+                return verifyQueryResultType(types.get(0), index);
+            } else {
+                for (Type type : types) {
+                    verifyQueryResultType(type, index);
+                }
+            }
+        } else {
+            final ClassInfo typeClass = index.getClassByName(t.name());
+            if (typeClass == null) {
+                throw new IllegalStateException(t.name() + " was not part of the Quarkus index");
+            }
+        }
+        return t;
+    }
+
+    protected DotName createSimpleInterfaceImpl(DotName ifaceName) {
+        String fullName = ifaceName.toString();
+
+        // package name: must be in the same package as the interface
+        final int index = fullName.lastIndexOf('.');
+        String packageName = "";
+        if (index > 0 && index < fullName.length() - 1) {
+            packageName = fullName.substring(0, index) + ".";
+        }
+
+        return DotName.createSimple(packageName
+                + (ifaceName.isInner() ? ifaceName.local() : ifaceName.withoutPackagePrefix()) + "_"
+                + HashUtil.sha1(ifaceName.toString()));
+    }
+
+    protected DotName getPrimitiveTypeName(DotName returnTypeName) {
+        if (DotNames.LONG.equals(returnTypeName)) {
+            return DotNames.PRIMITIVE_LONG;
+        }
+        if (DotNames.INTEGER.equals(returnTypeName)) {
+            return DotNames.PRIMITIVE_INTEGER;
+        }
+        if (DotNames.BOOLEAN.equals(returnTypeName)) {
+            return DotNames.PRIMITIVE_BOOLEAN;
+        }
+        return returnTypeName;
     }
 }

--- a/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/CustomQueryMethodsAdder.java
+++ b/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/CustomQueryMethodsAdder.java
@@ -24,7 +24,6 @@ import org.jboss.jandex.DotName;
 import org.jboss.jandex.IndexView;
 import org.jboss.jandex.MethodInfo;
 import org.jboss.jandex.Type;
-import org.jboss.jandex.Type.Kind;
 
 import io.quarkus.deployment.bean.JavaBeanUtil;
 import io.quarkus.gizmo.ClassCreator;
@@ -38,7 +37,6 @@ import io.quarkus.hibernate.orm.panache.common.runtime.AbstractJpaOperations;
 import io.quarkus.hibernate.orm.panache.runtime.AdditionalJpaOperations;
 import io.quarkus.panache.common.Parameters;
 import io.quarkus.panache.common.deployment.TypeBundle;
-import io.quarkus.runtime.util.HashUtil;
 import io.quarkus.spring.data.deployment.DotNames;
 import io.quarkus.spring.data.deployment.MethodNameParser;
 import io.quarkus.spring.data.runtime.TypesConverter;
@@ -275,7 +273,7 @@ public class CustomQueryMethodsAdder extends AbstractMethodsAdder {
 
                     // Find the type of data used in the result
                     // e.g. method.returnType() is a List that may contain non-entity elements
-                    Type resultType = verifyQueryResultType(method.returnType());
+                    Type resultType = verifyQueryResultType(method.returnType(), index);
                     DotName customResultTypeName = resultType.name();
 
                     if (customResultTypeName.equals(entityClassInfo.name())
@@ -291,7 +289,7 @@ public class CustomQueryMethodsAdder extends AbstractMethodsAdder {
                         if (Modifier.isInterface(resultClassInfo.flags())) {
                             // Find the implementation name, and use that for subsequent query result generation
                             customResultTypeName = customResultTypeNames.computeIfAbsent(customResultTypeName,
-                                    k -> createImplDotName(resultType.name()));
+                                    k -> createSimpleInterfaceImpl(resultType.name()));
 
                             // Remember the parameters for this usage of the custom type, we'll deal with it later
                             customResultTypes.computeIfAbsent(customResultTypeName,
@@ -333,7 +331,8 @@ public class CustomQueryMethodsAdder extends AbstractMethodsAdder {
                     }
 
                     generateFindQueryResultHandling(methodCreator, panacheQuery, pageableParameterIndex, repositoryClassInfo,
-                            entityClassInfo, methodReturnTypeDotName, null, method.name(), customResultTypeName);
+                            entityClassInfo, methodReturnTypeDotName, null, method.name(), customResultTypeName,
+                            Object[].class.getName());
                 }
 
             }
@@ -404,35 +403,6 @@ public class CustomQueryMethodsAdder extends AbstractMethodsAdder {
         return sort;
     }
 
-    // Make sure the return type is referencing a class we have indexed 
-    // somewhere along the way (e.g. non-entity return types in @Query methods)
-    // Unless it is some kind of collection containing multiple types, 
-    // return the type used in the query result.
-    private Type verifyQueryResultType(Type t) {
-        if (isHibernateSupportedReturnType(t.name())) {
-            return t;
-        }
-        if (t.kind() == Kind.ARRAY) {
-            return verifyQueryResultType(t.asArrayType().component());
-        } else if (t.kind() == Kind.PARAMETERIZED_TYPE) {
-            List<Type> list = t.asParameterizedType().arguments();
-            if (list.size() == 1) {
-                return verifyQueryResultType(list.get(0));
-            } else {
-                for (Type x : list) {
-                    verifyQueryResultType(x);
-                }
-                return t;
-            }
-        } else {
-            ClassInfo typeClassInfo = index.getClassByName(t.name());
-            if (typeClassInfo == null) {
-                throw new IllegalStateException(t.name() + " was not part of the Quarkus index");
-            }
-        }
-        return t;
-    }
-
     private List<String> getFieldNames(String queryString) {
         Matcher matcher = SELECT_CLAUSE.matcher(queryString);
         if (matcher.find()) {
@@ -457,21 +427,6 @@ public class CustomQueryMethodsAdder extends AbstractMethodsAdder {
         }
 
         return Collections.emptyList();
-    }
-
-    private DotName createImplDotName(DotName ifaceName) {
-        String fullName = ifaceName.toString();
-
-        // package name: must be in the same package as the interface
-        final int index = fullName.lastIndexOf('.');
-        String packageName = "";
-        if (index > 0 && index < fullName.length() - 1) {
-            packageName = fullName.substring(0, index) + ".";
-        }
-
-        return DotName.createSimple(packageName
-                + (ifaceName.isInner() ? ifaceName.local() : ifaceName.withoutPackagePrefix()) + "_"
-                + HashUtil.sha1(ifaceName.toString()));
     }
 
     private void generateCustomResultTypes(DotName interfaceName, DotName implName, Map<String, List<String>> queryMethods) {
@@ -552,18 +507,4 @@ public class CustomQueryMethodsAdder extends AbstractMethodsAdder {
         }
         return resultHandle;
     }
-
-    private DotName getPrimitiveTypeName(DotName returnTypeName) {
-        if (DotNames.LONG.equals(returnTypeName)) {
-            return DotNames.PRIMITIVE_LONG;
-        }
-        if (DotNames.INTEGER.equals(returnTypeName)) {
-            return DotNames.PRIMITIVE_INTEGER;
-        }
-        if (DotNames.BOOLEAN.equals(returnTypeName)) {
-            return DotNames.PRIMITIVE_BOOLEAN;
-        }
-        return returnTypeName;
-    }
-
 }

--- a/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/DerivedMethodsAdder.java
+++ b/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/DerivedMethodsAdder.java
@@ -4,7 +4,10 @@ import static io.quarkus.gizmo.FieldDescriptor.of;
 
 import java.lang.reflect.Modifier;
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
 
 import javax.transaction.Transactional;
 
@@ -17,7 +20,9 @@ import org.jboss.jandex.Type;
 import org.springframework.data.domain.Pageable;
 import org.springframework.data.domain.Sort;
 
+import io.quarkus.deployment.bean.JavaBeanUtil;
 import io.quarkus.gizmo.ClassCreator;
+import io.quarkus.gizmo.ClassOutput;
 import io.quarkus.gizmo.FieldDescriptor;
 import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.MethodDescriptor;
@@ -35,17 +40,27 @@ public class DerivedMethodsAdder extends AbstractMethodsAdder {
     private final IndexView index;
     private final String operationsName;
     private final FieldDescriptor operationsField;
+    private final ClassOutput nonBeansClassOutput;
+    private final Consumer<String> projectionClassCreatedCallback;
 
-    public DerivedMethodsAdder(IndexView index, TypeBundle typeBundle) {
+    public DerivedMethodsAdder(IndexView index, TypeBundle typeBundle, ClassOutput nonBeansClassOutput,
+            Consumer<String> projectionClassCreatedCallback) {
         this.index = index;
         operationsName = typeBundle.operations().dotName().toString();
         operationsField = of(operationsName, "INSTANCE", operationsName);
+        this.nonBeansClassOutput = nonBeansClassOutput;
+        this.projectionClassCreatedCallback = projectionClassCreatedCallback;
     }
 
     public void add(ClassCreator classCreator, FieldDescriptor entityClassFieldDescriptor,
             String generatedClassName, ClassInfo repositoryClassInfo, ClassInfo entityClassInfo) {
         MethodNameParser methodNameParser = new MethodNameParser(entityClassInfo, index);
         List<MethodInfo> repoMethods = new ArrayList<>(repositoryClassInfo.methods());
+
+        // Remember custom return type methods: {resultType:[methodName]}
+        Map<DotName, List<String>> customResultTypes = new HashMap<>(3);
+        Map<DotName, DotName> customResultTypeImplNames = new HashMap<>(3);
+
         //As intermediate interfaces are supported for spring data repositories, we need to search the methods declared in such interfaced and add them to the methods to implement list
         for (DotName extendedInterface : repositoryClassInfo.interfaceNames()) {
             if (GenerationUtil.isIntermediateRepository(extendedInterface, index)) {
@@ -149,8 +164,34 @@ public class DerivedMethodsAdder extends AbstractMethodsAdder {
                             methodCreator.readInstanceField(entityClassFieldDescriptor, methodCreator.getThis()),
                             methodCreator.load(finalQuery), sort, paramsArray);
 
+                    Type resultType = verifyQueryResultType(method.returnType(), index);
+                    DotName customResultTypeName = resultType.name();
+
+                    if (customResultTypeName.equals(entityClassInfo.name())
+                            || isHibernateSupportedReturnType(customResultTypeName)) {
+                        // no special handling needed
+                        customResultTypeName = null;
+                    } else {
+                        // If the custom type is an interface, we need to generate the implementation
+                        ClassInfo resultClassInfo = index.getClassByName(customResultTypeName);
+                        if (Modifier.isInterface(resultClassInfo.flags())) {
+                            // Find the implementation name, and use that for subsequent query result generation
+                            customResultTypeName = customResultTypeImplNames.computeIfAbsent(customResultTypeName,
+                                    k -> createSimpleInterfaceImpl(resultType.name()));
+
+                            // Remember the parameters for this usage of the custom type, we'll deal with it later
+                            customResultTypes.computeIfAbsent(customResultTypeName,
+                                    k -> new ArrayList<>()).add(method.name());
+                        } else {
+                            throw new IllegalArgumentException(
+                                    method.name() + " of Repository " + repositoryClassInfo
+                                            + " can only use interfaces to map results to non-entity types.");
+                        }
+                    }
+
                     generateFindQueryResultHandling(methodCreator, panacheQuery, pageableParameterIndex, repositoryClassInfo,
-                            entityClassInfo, returnType.name(), parseResult.getTopCount(), method.name(), null);
+                            entityClassInfo, returnType.name(), parseResult.getTopCount(), method.name(), customResultTypeName,
+                            entityClassInfo.name().toString());
 
                 } else if (parseResult.getQueryType() == MethodNameParser.QueryType.COUNT) {
                     if (!DotNames.PRIMITIVE_LONG.equals(returnType.name()) && !DotNames.LONG.equals(returnType.name())) {
@@ -233,5 +274,81 @@ public class DerivedMethodsAdder extends AbstractMethodsAdder {
                 }
             }
         }
+        for (Map.Entry<DotName, DotName> mapping : customResultTypeImplNames.entrySet()) {
+            DotName interfaceName = mapping.getKey();
+            DotName implName = mapping.getValue();
+            generateCustomResultTypes(interfaceName, implName, entityClassInfo, customResultTypes.get(implName));
+            projectionClassCreatedCallback.accept(implName.toString());
+        }
+    }
+
+    private void generateCustomResultTypes(DotName interfaceName, DotName implName, ClassInfo entityClassInfo,
+            List<String> queryMethods) {
+
+        ClassInfo interfaceInfo = index.getClassByName(interfaceName);
+
+        try (ClassCreator implClassCreator = ClassCreator.builder().classOutput(nonBeansClassOutput)
+                .interfaces(interfaceName.toString()).className(implName.toString())
+                .build()) {
+
+            Map<String, FieldDescriptor> fields = new HashMap<>(3);
+
+            for (MethodInfo method : interfaceInfo.methods()) {
+                String getterName = method.name();
+                String propertyName = JavaBeanUtil.getPropertyNameFromGetter(getterName);
+
+                Type returnType = method.returnType();
+                if (returnType.kind() == Type.Kind.VOID) {
+                    throw new IllegalArgumentException("Method " + method.name() + " of interface " + interfaceName
+                            + " is not a getter method since it returns void");
+                }
+                DotName fieldTypeName = getPrimitiveTypeName(returnType.name());
+
+                FieldDescriptor field = implClassCreator.getFieldCreator(propertyName, fieldTypeName.toString())
+                        .getFieldDescriptor();
+
+                // create getter (based on the interface)
+                try (MethodCreator getter = implClassCreator.getMethodCreator(getterName, returnType.toString())) {
+                    getter.setModifiers(Modifier.PUBLIC);
+                    getter.returnValue(getter.readInstanceField(field, getter.getThis()));
+                }
+
+                fields.put(getterName, field);
+            }
+
+            // Add static methods to convert from Object[] to this type
+            for (String queryMethod : queryMethods) {
+                try (MethodCreator convert = implClassCreator.getMethodCreator("convert_" + queryMethod,
+                        implName.toString(), entityClassInfo.name().toString())) {
+                    convert.setModifiers(Modifier.STATIC);
+
+                    ResultHandle newObject = convert.newInstance(MethodDescriptor.ofConstructor(implName.toString()));
+
+                    ResultHandle entity = convert.getMethodParam(0);
+                    final List<MethodInfo> availableMethods = entityClassInfo.methods();
+                    for (Map.Entry<String, FieldDescriptor> field : fields.entrySet()) {
+                        if (!getterExists(availableMethods, field.getKey())) {
+                            throw new IllegalArgumentException(field.getKey() + " method does not exists in "
+                                    + entityClassInfo.name().toString() + " class.");
+                        }
+
+                        FieldDescriptor f = field.getValue();
+                        convert.writeInstanceField(f, newObject, convert.invokeVirtualMethod(
+                                MethodDescriptor.ofMethod(entityClassInfo.name().toString(), field.getKey(), f.getType()),
+                                entity));
+                    }
+                    convert.returnValue(newObject);
+                }
+            }
+        }
+    }
+
+    private boolean getterExists(List<MethodInfo> methods, String getterName) {
+        for (MethodInfo method : methods) {
+            if (method.name().equals(getterName)) {
+                return true;
+            }
+        }
+        return false;
     }
 }

--- a/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/SpringDataRepositoryCreator.java
+++ b/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/generate/SpringDataRepositoryCreator.java
@@ -47,7 +47,7 @@ public class SpringDataRepositoryCreator {
         this.index = index;
         this.fragmentMethodsAdder = new FragmentMethodsAdder(fragmentImplClassResolvedCallback, index);
         this.stockMethodsAdder = new StockMethodsAdder(index, typeBundle);
-        this.derivedMethodsAdder = new DerivedMethodsAdder(index, typeBundle);
+        this.derivedMethodsAdder = new DerivedMethodsAdder(index, typeBundle, otherClassOutput, customClassCreatedCallback);
 
         // custom queries may generate non-bean classes
         this.customQueryMethodsAdder = new CustomQueryMethodsAdder(index, otherClassOutput, customClassCreatedCallback,

--- a/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/MovieRepository.java
+++ b/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/MovieRepository.java
@@ -68,6 +68,9 @@ public interface MovieRepository extends CrudRepository<Movie, Long> {
     @Query("SELECT title, rating FROM Movie WHERE title = ?1")
     MovieRating findRatingByTitle(String title);
 
+    // issue 13050
+    List<MovieProjection> findTitleAndRatingByRating(String rating);
+
     interface MovieCountByRating {
         String getRating();
 
@@ -79,4 +82,11 @@ public interface MovieRepository extends CrudRepository<Movie, Long> {
 
         String getRating();
     }
+
+    interface MovieProjection {
+        String getTitle();
+
+        String getRating();
+    }
+
 }

--- a/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/MovieResource.java
+++ b/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/MovieResource.java
@@ -137,7 +137,6 @@ public class MovieResource {
     @Produces("application/json")
     public Optional<MovieRepository.MovieRating> optionalTitleRating(@PathParam("title") String title) {
         Optional result = movieRepository.findOptionalRatingByTitle(title);
-        System.out.println(result);
         return result;
     }
 
@@ -158,4 +157,12 @@ public class MovieResource {
         movie.setDuration(1000);
         return movieRepository.save(movie);
     }
+
+    @GET
+    @Path("/titles/rating/{rating}")
+    @Produces("application/json")
+    public List<MovieRepository.MovieProjection> getTitlesByRating(@PathParam("rating") String rating) {
+        return movieRepository.findTitleAndRatingByRating(rating);
+    }
+
 }

--- a/integration-tests/spring-data-jpa/src/test/java/io/quarkus/it/spring/data/jpa/MovieResourceTest.java
+++ b/integration-tests/spring-data-jpa/src/test/java/io/quarkus/it/spring/data/jpa/MovieResourceTest.java
@@ -236,4 +236,11 @@ public class MovieResourceTest {
                 .statusCode(200)
                 .body(is("1"));
     }
+
+    @Test
+    void getTitle() {
+        when().get("/movie/titles/rating/PG-13").then()
+                .statusCode(200)
+                .body(containsString("Godzilla"));
+    }
 }

